src/main/java/com/google/sps/model/GmailResponse.java

```diff
@@ -0,0 +1,288 @@
+                    return message.getInternalDate() >= timeCutoffEpochMs;
+                  }
+                  throw new GmailMessageFormatException(
+                      "Messages must be of format MINIMAL, METADATA, or FULL");
+                })
+            .count();
+  }
+
+  /**
+   * Get the amount of unread, important emails from the last nDays given a list of user's unread
+   * emails. Emails without labelIds are skipped (they may be of the correct format, but there are
+   * no label ids, which results in a null return from google for labelIds)
+   *
+   * @param unreadEmails a list of Message objects from user's gmail account. Messages must be
+   *     either FULL, METADATA, or MINIMAL MessageFormat
+   * @return number of unread, important emails from last nDays days
+   */
+  private int findImportantUnreadEmailCount(List<Message> unreadEmails) {
+    return (int)
+        unreadEmails.stream()
+            .filter(
+                (message) ->
+                    message.getLabelIds() != null && message.getLabelIds().contains("IMPORTANT"))
+            .count();
+  }
+
+  /**
+   * Get the sender of the most unread emails from the last nDays days in a user's Gmail account
+   *
+   * @param unreadEmails a list of unread emails (with METADATA OR FULL MessageFormat)
+   * @return sender of the most unread emails from the last nDays days
+   * @throws GmailMessageFormatException if the messages do not contain an internal date and/or
+   *     headers (and are thus the wrong format)
+   */
+  private String findMostFrequentSender(List<Message> unreadEmails)
+      throws GmailMessageFormatException {
+    // This data structure is used in the hashmap, to associate two values (how many times
+    // the sender sent emails in the last nDays, and when their most recent sent email was) with
+    // the sender's email in the hashmap.
+    class EmailFrequencyWithDate {
+      private int timesSent;
+      private long mostRecentEmailTimestamp;
+
+      /**
+       * Create a new EmailFrequencyWithDate instance
+       *
+       * @param timestamp timestamp of email sent in milliseconds since epoch
+       */
+      public EmailFrequencyWithDate(long timestamp) {
+        timesSent = 1;
+        this.mostRecentEmailTimestamp = timestamp;
+      }
+
+      public int getTimesSent() {
+        return timesSent;
+      }
+
+      public long getMostRecentEmailTimestamp() {
+        return mostRecentEmailTimestamp;
+      }
+
+      /**
+       * Updates the mostRecentEmailTimestamp if the passed timestamp is later than the currently
+       * stored one.
+       *
+       * @param timestamp timestamp of email sent in milliseconds since epoch
+       */
+      public void updateMostRecentEmailTimestamp(long timestamp) {
+        if (timestamp > mostRecentEmailTimestamp) {
+          this.mostRecentEmailTimestamp = timestamp;
+        }
+      }
+
+      public void incrementTimesSent() {
+        timesSent++;
+      }
+    }
```

Comment:
Although I agree that this class only makes sense inside the method, I think class definitions inside methods hurts readability considerably. I'd rather define this inside `GmailResponse`.

Reply:
I ended up getting rid of this class. However, I used a new comparator class which I did declare at the class level inside `GmailResponseUtility`.

I completely agree with you - the readability gain is worth it, especially since the class is private.